### ISSUE TITLE
How can I tune the parameters to use the data from Kaggle?

### ISSUE THREAD

**Issue by artiiicy** [arcadelab/deepdrr#25]
Hi, I want to generate some DRR dataset from the CT volumetic images.

I got some CT data from [Kaggle](https://www.kaggle.com/andrewmvd/mosmed-covid19-ct-scans).
However, I couldn't find any information about those data that is required to tune the parameters from the codes.
(e.g. sensor_size, pixel_size, isocenter_distance etc..)

How can I set the parameters with the data from Kaggle or other datasets which are achieved from the opensource?

**Comment by mathiasunberath** [comment]
DeepDRR allows you to simulate digital radiography, as acquired by a [C-arm](https://www.researchgate.net/profile/Larry_Howell4/publication/301466164/figure/fig14/AS:667936964485138@1536259990578/The-GE-Healthcare-C-arm-used-for-imaging-in-the-operating-room-is-the-second-design.png) or digital radiography system, from CT scans (that in your case you got from Kaggle). 

Thus, in order to use DeepDRR, you will need to specify the camera parameters of the radiography system which follows approximately a [pinhole camera model](https://en.wikipedia.org/wiki/Pinhole_camera_model). To generate images you will need a [3 x 4 projection matrix](https://homepages.inf.ed.ac.uk/rbf/CVonline/LOCAL_COPIES/EPSRC_SSAZ/node3.html) that will map homogeneous 3D coordinates from your volume to 2D pixel coordinates in the detector, your resulting digitally reconstructed radiograph. 

I'd suggest you read up on those concepts to get started and we can then see where you stand. There are some other common issues you may run into that have been answered on here before.

**Comment by artiiicy** [comment]
Thanks for your comment.
I read all those concepts and appreciate to your suggestion. It was really beneficial.

The data that I got from Kaggle has the dimension of (512, 512, 54) and the size of (0.92, 0.92, 9.00) mm.
I can't find any other information.
How can I get the projection matrix matching those data.
Also, which codes need to be fixed?
Sorry for the question is still ambiguous.
Thanks.

**Comment by mathiasunberath** [comment]
Can you confirm that you have attempted to run our [usage example](https://github.com/arcadelab/deepdrr#usage) in the readme by loading your data and projecting it? 
You will probably need to adjust the center point (which should be set to the center of your CT, in most cases) and then you can check in greater detail the other components that go into the projection part of the code.

**Comment by artiiicy** [comment]
I appreciate your comment. I attempted to run the example codes with setting of the center below.
`center = volume.world_from_ijk @ deepdrr.geo.point(256, 256, 27)`
But I got this Error.

> Traceback (most recent call last):
>  File "/home/minjoon/Desktop/JupyterLab/deepDRR/reconstruction.py", line 18, in <module>
>    carm = deepdrr.CArm(isocenter=center)
> TypeError: __init__() missing 1 required positional argument: 'isocenter_distance'

I can fix it by `carm = deepdrr.CArm(isocenter=center, isocenter_distance = ?)`.
But I don't know what value to put for ?.
The below picture is my result when I put '0' to isocenter_distance.
![image](https://user-images.githubusercontent.com/48172859/106839388-731eb880-66e1-11eb-83cc-a37741bcb108.png)

**Comment by mathiasunberath** [comment]
OK, this looks like you are making progress. 0 as the isocenter distance is not a great because this essentially means that you are rotating around the X-ray source. May work but will make geometry debugging hard. Set it to about half the source-detector distance, in this case ~600. 
Then, you can play around a bit with the center. You should think a bit about what you would expect your images to look like which will help you see whether things are read in properly.

**Comment by artiiicy** [comment]
Hi, I'm sorry for asking a poor question that I adjusted the isocenter_distance to '0'. I fixed it 600.

I've tried to get valuable results. However, I still stuck in the same issue.

First, I guess my setting of the center point is wrong.
Here is my CT data.
![image](https://user-images.githubusercontent.com/48172859/107215140-44407380-6a4e-11eb-8aea-92acc79b425b.png)
As you can see it, dimension of the CT is [512, 512, 36].  So I adjusted the center like this.
`center = volume.world_from_ijk @ deepdrr.geo.point(256, 256, 18)`
Did I misunderstand what you mentioned about the adjustment of the center?

These are my codes and the results respectively.
![image](https://user-images.githubusercontent.com/48172859/107215876-4820c580-6a4f-11eb-9f44-63bcabbb50a3.png)
![image](https://user-images.githubusercontent.com/48172859/107215892-4fe06a00-6a4f-11eb-849e-bf20395efcc6.png)

When I set the center to (256, 256, 240), I got this results.
![image](https://user-images.githubusercontent.com/48172859/107216013-76060a00-6a4f-11eb-9146-d86ff847f35f.png)

What should I do to take off from this issue?

**Comment by mathiasunberath** [comment]
Can you generate a couple more images by rotating your camera source around the iso center in both angular directions? This will allow you to understand better from which direction you are looking at your volume.

**Comment by artiiicy** [comment]
Thanks for your comment.
By following your comment, I got this result when I set the `center = (256, 256, 200) source to detector dist = 1200, isocenter dist = 600`.

![image](https://user-images.githubusercontent.com/48172859/107365240-d95f6d00-6b1f-11eb-9cee-d40e6d8f0717.png)

I think It's a projection of the CT from axial view.
Following figures are some slices of the input.

![image](https://user-images.githubusercontent.com/48172859/107365970-e892ea80-6b20-11eb-9436-c65edcd67af6.png)
![image](https://user-images.githubusercontent.com/48172859/107366095-11b37b00-6b21-11eb-99e9-ba3b12059205.png)


It seems that the summation of the red lines makes the first figure.
But I can't find any 'bone-likely' structures, only black shadow.
I guess it's because the range is small, so the contrast doesn't be seen.
If I'm right, can I adjust the range of the results?

Here is my codes that gave me the result.
`import deepdrr
from scipy import io
import matplotlib.pyplot as plt
from pathlib import Path
import logging

from deepdrr import utils
from deepdrr import Volume, CArm, Projector
from deepdrr import geo

logger = logging.getLogger(__name__)

# volume = deepdrr.Volume.from_nifti('/path/to/ct_image.nii.gz')
volume = deepdrr.Volume.from_nifti('./data/study_1109.nii.gz')

camera_intrinsics = deepdrr.geo.CameraIntrinsicTransform.from_sizes(
    sensor_size=512,
    pixel_size=0.78,
    source_to_detector_distance=1200,
)


center = volume.world_from_ijk @ deepdrr.geo.point(256, 256, 200)

carm = deepdrr.CArm(isocenter=center, isocenter_distance = 600)

# Angles to take projections over
min_theta = 180
max_theta = 240
min_phi = 60
max_phi = 180
spacing_theta = 30
spacing_phi = 30

with deepdrr.Projector(volume, camera_intrinsics, carm) as projector:
    images = projector.project_over_carm_range(
            (min_phi, max_phi, spacing_phi),
            (min_theta, max_theta, spacing_theta)
    )
    
    
# get the thetas an phis for file names
phis, thetas = utils.generate_uniform_angles(
    (min_phi, max_phi, spacing_phi),
    (min_theta, max_theta, spacing_theta)
)
    
output_dir = Path(f'results')
output_dir.mkdir(exist_ok=True)

for i, image in enumerate(images):
    plt.imshow(image, cmap="gray")
    plt.title(f'phi, theta = {phis[i], thetas[i]}')
    output_path = output_dir / f'image_phi={int(phis[i])}_theta={int(thetas[i])}.png'
    logger.info(f'writing image {output_path}')
    plt.savefig(output_path)`

**Comment by mathiasunberath** [comment]
So the problem I see is that you are looking at your volume from the head-foot direction. This is not usually a view that is possible with X-ray machines. You should rotate around 90 degrees so that you get an anterior-posterior view of your CT scan. Once you have that geometric configuration you should start seeing some structure.

**Comment by artiiicy** [comment]
I followed your comment, but I can't see any structure.

![image](https://user-images.githubusercontent.com/48172859/107732335-72b69b00-6d3b-11eb-8cd7-4d1a0e3bb26a.png)
![image](https://user-images.githubusercontent.com/48172859/107732352-7d713000-6d3b-11eb-946f-3509f548604c.png)
![image](https://user-images.githubusercontent.com/48172859/107732404-9a0d6800-6d3b-11eb-99be-5bd98c860184.png)
![image](https://user-images.githubusercontent.com/48172859/107732422-a691c080-6d3b-11eb-80e4-bdbdc193da14.png)

The other setting of the parameters are same with picture above.
Am I doing something wrong? :(

**Comment by mathiasunberath** [comment]
I believe to see lung structures in your first images which is good. This suggests that right now your images are simply too dark. This can have multiple reasons:
1) your primary (the number of photons) are not sufficiently high so you don't get enough signal at the detector.
2) your dicoms are not properly segmented into the different materials. This can have several reasons, for example that intensity values are not read properly or the V-net segmentation does not work. Here I would recommend to use threshold based material decomposition and manually check the result. 
3) your dicom is not read properly so that all intensity values are too high. 
4) some combination of the above
5) something else.

**Comment by artiiicy** [comment]
I do appreciate for your comment again.

I checked the results step by step.
First, I plotted the results of segmentation and the result of mass_attenuation.
These are the results.
(From left to right: bone, soft tissue, air, attenuation)

1. Coronal view
![image](https://user-images.githubusercontent.com/48172859/107948823-b3cadb80-6fd7-11eb-9511-1e4b96e090f5.png)

2. Sagittal view
![image](https://user-images.githubusercontent.com/48172859/107948564-62bae780-6fd7-11eb-8895-6d3d9e518d65.png)

I think that attenuation part is something wrong.

I am wondering if I can get data you worked with. I just want to see the valuable results. :(

**Comment by mathiasunberath** [comment]
The individual images look good. We had an issue with the version that @benjamindkilleen fixed. Maybe you are running into that one if your version is that specific one. Maybe you can update and @benjamindkilleen can provide some guidance on that specific bug?

**Comment by benjamindkilleen** [comment]
@artiiicy can you try installing the v1.0.1-alpha branch from source? This is an unstable branch, but it is the most up to date as of now and may have a fix to your problem.

**Comment by artiiicy** [comment]
Thanks for your help.
Unfortunately, v.1.0.1-alpha shows the same results.

I checked my input data's intensity and it has the range about [-1024 ~ 3071].
Is there any possibility that my input data is invalid?

I used the nii data which is achieved from conversion of the below data.
https://data.idoimaging.com/dicom/1020_abdomen_ct/1020_abdomen_ct_510_jpg.zip

**Comment by benjamindkilleen** [comment]
By the data's intensity, do you mean the CT data? From that range, you need to convert from Hounsfield units to density. Can you post a code snippet for how you are instantiating the Volume object? You may need to use the `from_hu` or another classmethod rather than the `Volume()` class constructor directly.

**Comment by artiiicy** [comment]
Yes, The range about [-1024 ~ 3071] means the intensity of CT data.

Here is my code snippet for loading the Volume object.
As you can see, I just used the `volume()` class.
![image](https://user-images.githubusercontent.com/48172859/108069040-3fab3900-70a6-11eb-8c33-095bfe8779d3.png)

Is there any written codes in your repository for conversion from HU to density?
(I can't find the `from_hu` class.)

**Comment by benjamindkilleen** [comment]
On the alpha branch, `from_hu` is a classmethod like `from_nifti` that takes in the HU CT volume as a numpy array, rather than a path to a file. The `load_dicom` module contains the code for converting HU to density. However, from your filename, is your nifti file already converted to density? Or is that referring to the conversion from DICOM to NifTi? If you examine the values of the NifTi file directly, what are they?

**Comment by artiiicy** [comment]
Oh, Sorry. I searched on the master not branch.

My filename means the conversion from DICOM to NifTi.
As I mentioned, I used this data (https://data.idoimaging.com/dicom/1020_abdomen_ct/1020_abdomen_ct_510_jpg.zip)
which is DICOM.
So I converted it to NifTi and used as input.

The values of the NifTi file are [-1024 ~ 3071].
Sorry to make you confused.

Additionally, I had critical mistakes. I used the codes of the master, not branch.
After I used the codes of `v.1.0.1-alpha`, I got this DRR results.
![image](https://user-images.githubusercontent.com/48172859/108203701-35e90a80-7166-11eb-8c2f-3d6cfd85794c.png)

I appreciate your help so much.
This is my code-snippet. (I don't use the `from_hu`. The `from_nifti` of the `v1.0.1-alpha` works well.)

![image](https://user-images.githubusercontent.com/48172859/108204461-2d450400-7167-11eb-8cfa-cabdaca9bcb8.png)
